twill-core/src/main/java/org/apache/twill/internal/AbstractTwillController.java

```diff
@@ -71,21 +71,37 @@
   private static final Logger LOG = LoggerFactory.getLogger(AbstractTwillController.class);
   private static final Gson GSON = new Gson();
 
+  private final String appName;
+  private final RunId runId;
   private final Queue<LogHandler> logHandlers;
   private final KafkaClientService kafkaClient;
   private ZKDiscoveryService discoveryServiceClient;
   private Cancellable logCancellable;
 
-  public AbstractTwillController(RunId runId, ZKClient zkClient, Iterable<LogHandler> logHandlers) {
+  public AbstractTwillController(String appName, RunId runId, ZKClient zkClient, boolean logCollectionEnabled,
+                                 Iterable<LogHandler> logHandlers) {
     super(runId, zkClient);
+    this.appName = appName;
+    this.runId = runId;
     this.logHandlers = new ConcurrentLinkedQueue<>();
-    this.kafkaClient = new ZKKafkaClientService(ZKClients.namespace(zkClient, "/" + runId.getId() + "/kafka"));
-    Iterables.addAll(this.logHandlers, logHandlers);
+
+    // When addressing TWILL-147, need to check if the given ZKClient is
+    // actually used by the Kafka used for log collection
+    if (logCollectionEnabled) {
+      this.kafkaClient = new ZKKafkaClientService(ZKClients.namespace(zkClient, "/" + runId.getId() + "/kafka"));
+      Iterables.addAll(this.logHandlers, logHandlers);
+    } else {
+      this.kafkaClient = null;
+      if (!Iterables.isEmpty(logHandlers)) {
+        LOG.warn("Log collection is disabled for application {} with runId {}. " +
+                   "Adding log handler won't get any logs.", appName, runId);
+      }
+    }
   }
 
   @Override
   protected synchronized void doStartUp() {
-    if (!logHandlers.isEmpty()) {
+    if (kafkaClient != null && !logHandlers.isEmpty()) {
```

Comment:
you mean have the logic as

```java
if (kafkaClient == null) {
  return;
}
if (!logHandlers.isEmpty()) {
  kafkaClient.startAndWait();
  logCancellable = kafkaClient.getConsumer().....
}
```

??
Doesn't look simpler than the existing code to me.